newrelic/api/transaction.py

```diff
@@ -1540,17 +1540,49 @@ def record_log_event(self, message, level=None, timestamp=None, attributes=None,
         timestamp = timestamp if timestamp is not None else time.time()
         level = str(level) if level is not None else "UNKNOWN"
 
-        if not message or message.isspace():
-            _logger.debug("record_log_event called where message was missing. No log event will be sent.")
+        # Unpack message and attributes from dict inputs
+        if isinstance(message, dict):
+            message_attributes = {k: v for k, v in message.items() if k != "message"}
+            message = str(message.get("message", ""))
+        else:
+            message_attributes = None
+
+        # Exit early for invalid message type after unpacking
+        is_string = isinstance(message, six.string_types)
+        if message is not None and not is_string:
+            _logger.debug("record_log_event called where message was not found. No log event will be sent.")
             return
 
-        message = truncate(message, MAX_LOG_MESSAGE_LENGTH)
+        # Exit early if no message or attributes found
+        no_message = not message or message.isspace()
+        if not attributes and not message_attributes and no_message:
+            _logger.debug("record_log_event called where message was missing, and no attributes found. No log event will be sent.")
+            return
+
+        # Truncate the now unpacked and string converted message
+        if is_string: 
+            message = truncate(message, MAX_LOG_MESSAGE_LENGTH)
+
+        # Collect attributes from linking metadata, context data, and message attributes
+        collected_attributes = {}
+        if settings and settings.application_logging.forwarding.context_data.enabled:
+            if attributes:
+                context_attributes = resolve_logging_context_attributes(attributes, settings.attribute_filter, "context.")
+                if context_attributes:
+                    collected_attributes.update(context_attributes)
+
+            if message_attributes:
+                message_attributes = resolve_logging_context_attributes(message_attributes, settings.attribute_filter, "message.")
+                if message_attributes:
+                    collected_attributes.update(message_attributes)
+
+            # Exit early if no message or attributes found after filtering
+            if not collected_attributes and no_message:
+                _logger.debug("record_log_event called where message was missing, and no attributes found. No log event will be sent.")
+                return
 
-        collected_attributes = get_linking_metadata()
-        if attributes and (settings and settings.application_logging.forwarding.context_data.enabled):
-            context_attributes = resolve_logging_context_attributes(attributes, settings.attribute_filter, "context.")
-            if context_attributes:
-                collected_attributes.update(context_attributes)
+        # Finally, add in linking attributes after checking that there is a valid message or at least 1 attribute
```

Comment:
What do you think about moving the above conditional:
```
            if no_message and not collected_attributes:
                _logger.debug("record_log_event called where message was missing, and no attributes found. No log event will be sent.")
                return
```
down here outside the `context_data.enabled` conditional and removing the conditional above that checks the no_message and attributes before they are filtered?

I could see a good reason for keeping these conditionals separate if the first one improved performance by exiting early before the filtering but I don't see that it results in a worthwhile performance improvement. If the attributes are not present, it doesn't try to filter them at all so I don't see the point in having two of these early exit checks. It seems like it'd be simpler to have just the one at the end.